[WIP] Computer vision util: vision visualizer

molbap opened the pull request:
# What does this PR do?

This PR is a draft to add a new utility tool to display in a notebook (or in ASCII if I crack it) how any processor from any model integrated in `transformers` sees an image. 

Specifically, it is about making readable at a glance how the `image_processor` transforms a given input in  `pixel_values`, and I'll see to add text as well maybe later.

Example usage, for SmolVLM:

```python
 from transformers.utils.processor_visualizer_utils import ImageVisualizer
from PIL import Image

viz = ImageVisualizer('HuggingFaceTB/SmolVLM2-2.2B-Instruct')

image_path = "hf-logo.png"
image = Image.open(image_path).resize((1000, 1349))
viz.visualize(image)
```
should get you the correct tiling + global image

![image](https://github.com/user-attachments/assets/d6bc1b6a-4271-4094-a169-9ce11576d596)

molbap committed:
draft the vision visualizer

ArthurZucker commented:
🤩👨‍🍳

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_36892). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

molbap committed:
Merge branch 'main' into vision_visualizer

molbap commented:
Sounds good @zucchini-nlp!

molbap committed:
Merge branch 'main' into vision_visualizer

molbap committed:
move processor visualizer

molbap committed:
fixup

molbap committed:
better namings

molbap committed:
add captions and better tiling detection

molbap committed:
fixup

molbap commented:
Writing down a lot these days, reopened and finished this on the way. cc @ArthurZucker and @zucchini-nlp if you want to take a look! Works for ~100 models or so, added the few that don't but took no action upon them now

Supports most image/Text models. Works for:

### 1. High patch counts

<img width="830" height="843" alt="image" src="https://github.com/user-attachments/assets/d7888c74-79c6-43ac-95fe-73937af78ac4" />

### 2. Low patch counts
<img width="933" height="1017" alt="image" src="https://github.com/user-attachments/assets/66c74948-734c-4007-ab2e-603418285eb4" />

### 3. Tiled output with global image (most recent/common)
<img width="470" height="971" alt="image" src="https://github.com/user-attachments/assets/5ef754c9-f9a5-4ba3-8988-bbc8157d55f5" />


### 4. Tiled output without global image (some llava models)

<img width="939" height="713" alt="image" src="https://github.com/user-attachments/assets/4683ced1-34cd-48d9-9274-6da78e37500f" />

Added as well the `default_processor_message` mentioned. I'll move on to the video processor later I think. 
```python
visualizer.default_message(full_output=True)
"""
For a 224x224 RGB png image: 

'<|im_start|>User:<fake_token_around_image><row_1_col_1><image>[...81 tokens...]<image><fake_token_around_image><row_1_col_2><image>[...81 tokens...]<image><fake_token_around_image><row_1_col_3><image>[...81 tokens...]<image><fake_token_around_image><row_1_col_4><image>[...81 tokens...]<image>\n<fake_token_around_image><row_2_col_1><image>[...81 tokens...]<image><fake_token_around_image><row_2_col_2><image>[...81 tokens...]<image><fake_token_around_image><row_2_col_3><image>[...81 tokens...]<image><fake_token_around_image><row_2_col_4><image>[...81 tokens...]<image>\n<fake_token_around_image><row_3_col_1><image>[...81 tokens...]<image><fake_token_around_image><row_3_col_2><image>[...81 tokens...]<image><fake_token_around_image><row_3_col_3><image>[...81 tokens...]<image><fake_token_around_image><row_3_col_4><image>[...81 tokens...]<image>\n<fake_token_around_image><row_4_col_1><image>[...81 tokens...]<image><fake_token_around_image><row_4_col_2><image>[...81 tokens...]<image><fake_token_around_image><row_4_col_3><image>[...81 tokens...]<image><fake_token_around_image><row_4_col_4><image>[...81 tokens...]<image>\n\n<fake_token_around_image><global-img><image>[...81 tokens...]<image><fake_token_around_image>Please describe this image.<end_of_utterance>\nAssistant:'
"""
```

I briefly looked into what would be a terminal/ASCII-friendly version but no luck! 

Hopefully that helps users and adds glance value to processors.

molbap committed:
Merge branch 'main' into vision_visualizer

zucchini-nlp commented on src/transformers/utils/processor_visualizer_utils.py:
```diff
@@ -0,0 +1,399 @@
+        self.channel_stds = getattr(image_processor, "image_std", [0.229, 0.224, 0.225])
+
+        # Image token marker used in prompts
+        self.image_token_marker = getattr(self.processor, "image_token_index", "<image>")
```
Ig it is `image_token_id` to cover more models, isn't it?

molbap replied:
ah, likely :D

zucchini-nlp commented on src/transformers/utils/processor_visualizer_utils.py:
```diff
@@ -0,0 +1,399 @@
+
+        return pixel_values
+
+    def _to_pil(self, image_input: Union[str, np.ndarray, Image.Image]) -> Image.Image:
```
not sure if we want the viz to be isolated from transformers helpers. There is a ready `to_pil_image` in general transforms 

https://github.com/huggingface/transformers/blob/555cbf59178134d1b713a7022129aaddfe6e70cf/src/transformers/image_utils.py#L576

molbap replied:
yeah no point in isolating, I'll regroup!

zucchini-nlp commented on src/transformers/utils/processor_visualizer_utils.py:
```diff
@@ -0,0 +1,399 @@
+        else:
+            raise TypeError(f"Unsupported image input type: {type(image_input)}")
+
+    def _unnormalize(
```
Not related to this PR: would be supeer nice to move this to general transforms at some point. It is already used by a few image generation models

molbap replied:
moved it to image_transform, lmk

zucchini-nlp commented on src/transformers/utils/processor_visualizer_utils.py:
```diff
@@ -0,0 +1,399 @@
+
+            def compress_image_token_run(match: re.Match) -> str:
+                n_tokens = match.group(0).count(image_token_string)
+                return f"{image_token_string}[...{n_tokens} tokens...]{image_token_string}"
```
love the `n_tokens` compression pattern

Btw, sometimes we have a non-contiguous set of image tokens, separated by special row/col tokens. We can consider covering it in later PRs

zucchini-nlp reviewed (approved):
Very very nice and helpful for the community! Just left a few questions. Probably we can come back later and improve by supporting more arch

molbap commented:
The community comment is exactly the kind of things I want to cover yes! Will open a follow-up PR

merveenoyan commented:
@molbap I think this would make a great demo! would love to help you build or build it myself, however you want :)

molbap committed:
add an unnormalize image method

molbap committed:
use existing methods, add default image

zucchini-nlp commented on src/transformers/image_transforms.py:
```diff
@@ -452,6 +452,105 @@ def normalize(
+    # target layout: default to NHWC so downstream viz works out of the box
+    target_format = data_format or ChannelDimension.LAST
+
+    # collapse any leading batch dims into one, preserving (C,H,W) or (H,W,C)
+    if input_data_format == ChannelDimension.FIRST:
+        # layout: [*, C, H, W]
+        lead = int(np.prod(image.shape[: image.ndim - 3])) if image.ndim > 3 else 1
+        if image.ndim == 3:
+            c, h, w = image.shape
+            image = image.reshape(1, c, h, w)
+            lead = 1
+        else:
+            c, h, w = image.shape[-3:]
+        image = image.reshape(lead, c, h, w)
+        if target_format == ChannelDimension.LAST:
+            image = np.transpose(image, (0, 2, 3, 1))  # -> [N, H, W, C]
+    else:
+        # layout: [*, H, W, C]
+        lead = int(np.prod(image.shape[: image.ndim - 3])) if image.ndim > 3 else 1
+        if image.ndim == 3:
+            h, w, c = image.shape
+            image = image.reshape(1, h, w, c)
+            lead = 1
+        else:
+            h, w, c = image.shape[-3:]
+        image = image.reshape(lead, h, w, c)
+        if target_format == ChannelDimension.FIRST:
+            image = np.transpose(image, (0, 3, 1, 2))  # -> [N, C, H, W]
```
this part is specific for plotting only. Imo we should keep images shape intact and only cast to `data_format` at the end, similar to how other transforms are done

zucchini-nlp replied:
yeah, like a general transform fn. Looks like it is getting more common lately

zucchini-nlp commented on src/transformers/image_transforms.py:
```diff
@@ -452,6 +452,105 @@ def normalize(
+        std_arr = torch.as_tensor(std_seq, dtype=image.dtype, device=image.device).view(bshape)
+        image = image * std_arr + mean_arr
+
+    # convert to numpy for plotting
+    if is_torch:
+        image = image.detach().cpu().numpy()
+        is_np = True  # from here on
+
```
same here

zucchini-nlp commented on src/transformers/image_transforms.py:
```diff
@@ -452,6 +452,105 @@ def normalize(
+    bshape[ch_axis] = num_channels
+
+    if is_np:
+        mean_arr = np.asarray(mean_seq, dtype=image.dtype).reshape(bshape)
+        std_arr = np.asarray(std_seq, dtype=image.dtype).reshape(bshape)
```
nor sure if reshape is necessary. If we have broadcastable shapes, numpy handles the rest isn't it?

molbap replied:
I think I had weird shapes at some point here... but you're right, should not need to reshape

zucchini-nlp reviewed (commented):
Left a few comments 😉

molbap committed:
solve unequal cropping

molbap committed:
Merge branch 'main' into vision_visualizer

molbap committed:
biig update for ragged inputs why not

molbap committed:
fixup

molbap committed:
push working qwen vl viz

molbap commented:
Came back to this lol. cc @zucchini-nlp now it should also work for Qwen-type models with ragged tensors! (gives a biiiig grid though)
